protocols/gossipsub/src/handler.rs

```diff
@@ -495,60 +484,83 @@ impl ConnectionHandler for Handler {
-            return;
-        }
+                if event.is_inbound()
+                    && handler.inbound_substreams_created == MAX_SUBSTREAM_CREATION
+                {
+                    log::warn!(
+                        "The maximum number of inbound substreams created has been exceeded"
+                    );
+                    *self = Handler::Disabled(DisabledHandler::MaxStreamCreation);
+                    return;
+                }
 
-        if event.is_outbound() && self.outbound_substreams_requested == MAX_SUBSTREAM_CREATION {
-            // Too many outbound substreams have been created, disable the handler.
-            self.keep_alive = KeepAlive::No;
-            log::warn!("The maximum number of outbound substreams created has been exceeded");
-            return;
-        }
+                if event.is_outbound()
+                    && handler.outbound_substreams_requested == MAX_SUBSTREAM_CREATION
+                {
+                    log::warn!(
+                        "The maximum number of outbound substreams created has been exceeded"
+                    );
+                    *self = Handler::Disabled(DisabledHandler::MaxStreamCreation);
+                    return;
+                }
 
-        match event {
-            ConnectionEvent::FullyNegotiatedInbound(fully_negotiated_inbound) => {
-                self.on_fully_negotiated_inbound(fully_negotiated_inbound)
-            }
-            ConnectionEvent::FullyNegotiatedOutbound(fully_negotiated_outbound) => {
-                self.on_fully_negotiated_outbound(fully_negotiated_outbound)
-            }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError {
-                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
-                ..
-            }) => {
-                log::debug!("Dial upgrade error: Protocol negotiation timeout");
-            }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError {
-                error: ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)),
-                ..
-            }) => void::unreachable(e),
-            ConnectionEvent::DialUpgradeError(DialUpgradeError {
-                error:
-                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)),
-                ..
-            }) => {
-                // The protocol is not supported
-                self.protocol_unsupported = true;
-                log::debug!("The remote peer does not support gossipsub on this connection");
-            }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError {
-                error:
-                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(
-                        NegotiationError::ProtocolError(e),
-                    )),
-                ..
-            }) => {
-                log::debug!("Protocol negotiation failed: {e}")
+                match event {
```

Comment:
We started adding `transpose` functions to this event such that you can split it directly into an `Either` and can combine things like `Handler::Disabled` and the `unreachable` into one match arm. Might be worth doing here as well.

Reply:
Neat! But if I am not mistaken the `transpose` methods are `pub(crate)`. Was that a deliberate choice? If not, mind pushing a commit to this branch @thomaseizinger?

Reply:
Ah interesting. I don't see a reason why they couldn't be public as well. Feel free to make them public.

Reply:
`libp2p-swarm` implements `transpose` for `    FullyNegotiatedInbound<Either<_, _>>, Either<_, _>>`. Note the second `Either`. In this case both `ProtocolConfig` and `DeniedUpgrade` use a `()` for the `InboundUpgradeInfo`, thus the `libp2p-swarm` `transpose` does not apply.

@thomaseizinger I suggest we proceed here without this change. Fine by you?

Reply:
Yep, fine by me.